### PR TITLE
Fix issues related to focus management

### LINKED ISSUE #184: iron-list.html:1257 Uncaught RangeError: <item> not found
This started after the last update.
` <iron-list id="list" items="[[items]]" as="item" scroll-target="html" selection-enabled multi-selection>
        <template>
...
        </template>
      </iron-list>`

If you scroll up and make `this.set('items',[]);` the error will occur and the list will become unusable. I use the same list to show different things. It worked well till now. I easily reproduced it with the demo by scrolling up, then click a button that set items to [].
Thanks.

### PR DESCRIPTION
**Pull request by blasten**
Should fix #184, #186, #189, #192
cc @kevinpschaaf

### CODE REVIEW

**Comment by miztroh** [comment]
@frankiefu Can we get a new release?

**Comment by frankiefu** [comment]
We found couple issues could be related to this patch and we are looking at it.  Will do a release when the patch is ready.

File: iron-list.html [hunk 1]
```diff
@@ -1383,132 +1383,167 @@
...
-        if (force) {
-          this.scrollToIndex(idx);
-          return this._getPhysicalItemForIndex(idx, false);
-        }
-        return null;
-      }
-      var item = this._getNormalizedItem(idx);
-      var physicalItem = this._physicalItems[this._physicalIndexForKey[this._collection.getKey(item)]];
+    _isIndexVisible: function(idx) {
+      return this._focusedIndex >= this.firstVisibleIndex && this._focusedIndex <= this.lastVisibleIndex;
```
Comment (frankiefu) [inline iron-list.html]: s/this._focusedIndex/idx
```
return idx >= this.firstVisibleIndex && idx <= this.lastVisibleIndex;
```

Comment (blasten) [inline iron-list.html]: good catch

**Comment by frankiefu** [comment]
LGTM

**Pull request merged**